ISISMuon/MuonDataLib
# Write event data to muon nexus file

### Description of work
This adds the code needed to allow for some of the new muon event nexus files to be loaded and converted into a muon nexus histogram file that Wimda can read. 

#### Summary of work

I have abstracted and added some of the classes for the nexus file properties (e.g. raw_data). This reduces the duplication of code. 

I have also created a cache that allows event information to be easily shared between nexus file properties. The cache is limited to data that is relevant for nexus file properties. i.e. stores histograms instead of events. 




Fixes #14 . 


### To test:


Warning: If the cython code crashes no feedback is provided. This is because the error checking has been turned off to increase the speed. A work around is shown below. 

```
from MuonDataLib.data.loader.load_events import load_events
import os

file_name = 'HIFIRun0.nxs'
file = os.path.join(os.path.dirname(__file__),
                        '..',
                        file_name)

data = load_events(file, 64)
print('start')

output = 'HIFI0.nxs'

data.save_histograms(Output)
print('end')

```

If it works it save the file and then print the message "end". If you don't get the end message then it has crashed. 

This doesn't work for the simulated data that I have as it is missing some metadata. It should work on stuff coming from the beamline. 

#### Code Review

- Is the code of an acceptable quality?
- Are the unit tests small and test the class in isolation?
- Does the documentation build [ReadTheDocs](https://readthedocs.org/projects/muondatalib/builds/)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**.
